README.md
zweui.js
=====

[![npm version](https://img.shields.io/npm/v/weui.js.svg)](https://www.npmjs.com/package/zweui.js/)


### 概述

[WeUI](https://github.com/Tencent/weui.git) 的轻量级 js 封装。
[WeUI.js](https://github.com/Tencent/weui.js.git)的扩展。

注意：由于微信小程序不支持dom操作，所以weui.js并不适用于小程序。



### 手机预览

![https://zafir.cn/zweui.js/](https://zafir.cn/zweui.js/qr.png)

[https://zafir.cn/zweui.js/](https://zafir.cn/zweui.js/)


### 开发

安装

```shell
git clone https://github.com/zafirlee/zweui.js.git
cd zweui.js
npm install
npm start
```

编译

```shell
npm run build
```


### 使用

#### global 



#### import as module

```javascript
import 'zweui';
import zweui from 'zweui.js';

weui.alert('alert');
```

### 文档

[Documents](https://github.com/zafirlee/zweui.js/tree/master/docs/README.md)

### 贡献

如果你有好的意见或建议，欢迎给我们提issue或pull request。

### License
The MIT License(http://opensource.org/licenses/MIT)

请自由地享受和参与开源


CHANGELOG.md



#### v0.0.1 (2020-12-15)

初始发布

- Initial release


src/zweui.js
/*
* Tencent is pleased to support the open source community by making WeUI.js available.
*
* Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
*
* Licensed under the MIT License (the "License"); you may not use this file except in compliance
* with the License. You may obtain a copy of the License at
*
*       http://opensource.org/licenses/MIT
*
* Unless required by applicable law or agreed to in writing, software distributed under the License is
* distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
* either express or implied. See the License for the specific language governing permissions and
* limitations under the License.
*/

import dialog from './dialog/dialog';
import alert from './alert/alert';
import confirm from './confirm/confirm';
import toast from './toast/toast';
import loading from './loading/loading';
import actionSheet from './actionSheet/actionSheet';
import topTips from './topTips/topTips';
import searchBar from './searchBar/searchBar';
import tab from './tab/tab';
import form from './form/form';
import uploader from './uploader/uploader';
import {picker, datePicker, distPicker} from './picker/picker';
import gallery from './gallery/gallery';
import slider from './slider/slider';

export default {
    dialog,
    alert,
    confirm,
    toast,
    loading,
    actionSheet,
    topTips,
    searchBar,
    tab,
    form,
    uploader,
    picker,
    datePicker,
    distPicker,
    gallery,
    slider
};


src/picker/picker.js
/*
* Tencent is pleased to support the open source community by making WeUI.js available.
*
* Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
*
* Licensed under the MIT License (the "License"); you may not use this file except in compliance
* with the License. You may obtain a copy of the License at
*
*       http://opensource.org/licenses/MIT
*
* Unless required by applicable law or agreed to in writing, software distributed under the License is
* distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
* either express or implied. See the License for the specific language governing permissions and
* limitations under the License.
*/

import $ from '../util/util';
import cron from './cron';
import './scroll';
import './districts';
import * as util from './util';
import pickerTpl from './picker.html';
import groupTpl from './group.html';
import districts from './districts';

function Result(item) {
    if (typeof item != 'object') {
        item = {
            label: item,
            value: item
        };
    }
    $.extend(this, item);
}

Result.prototype.toString = function () {
    return this.value;
};
Result.prototype.valueOf = function () {
    return this.value;
};

let _sington;
let temp = {}; // temp 存在上一次滑动的位置

/**
 * picker 多列选择器。
 * @param {array} items picker的数据，即用于生成picker的数据，picker的层级可以自己定义，但建议最多三层。数据格式参考example。
 * @param {Object} options 配置项
 * @param {number=} [options.depth] picker深度(也就是picker有多少列) 取值为1-3。如果为空，则取items第一项的深度。
 * @param {string=} [options.id=default] 作为picker的唯一标识，作用是以id缓存当时的选择。（当你想每次传入的defaultValue都是不一样时，可以使用不同的id区分）
 * @param {string=} [options.className] 自定义类名
 * @param {string=} [options.title] 自定义标题
 * @param {string=} [options.desc] 自定义描述
 * @param {string=} [options.container] 指定容器
 * @param {array=} [options.defaultValue] 默认选项的value数组
 * @param {function=} [options.onChange] 在picker选中的值发生变化的时候回调
 * @param {function=} [options.onConfirm] 在点击"确定"之后的回调。回调返回选中的结果(Array)，数组长度依赖于picker的层级。
 * @param {function=} [options.onClose] picker关闭后的回调
 *
 * @example
 * // 单列picker
 * weui.picker([
 * {
 *     label: '飞机票',
 *     value: 0,
 *     disabled: true // 不可用
 * },
 * {
 *     label: '火车票',
 *     value: 1
 * },
 * {
 *     label: '汽车票',
 *     value: 3
 * },
 * {
 *     label: '公车票',
 *     value: 4,
 * }
 * ], {
 *    className: 'custom-classname',
 *    container: 'body',
 *    defaultValue: [3],
 *    onChange: function (result) {
 *        console.log(result)
 *    },
 *    onConfirm: function (result) {
 *        console.log(result)
 *    },
 *    id: 'singleLinePicker'
 * });
 *
 * @example
 * // 多列picker
 * weui.picker([
 *     {
 *         label: '1',
 *         value: '1'
 *     }, {
 *         label: '2',
 *         value: '2'
 *     }, {
 *         label: '3',
 *         value: '3'
 *     }
 * ], [
 *     {
 *         label: 'A',
 *         value: 'A'
 *     }, {
 *         label: 'B',
 *         value: 'B'
 *     }, {
 *         label: 'C',
 *         value: 'C'
 *     }
 * ], {
 *     defaultValue: ['3', 'A'],
 *     onChange: function (result) {
 *         console.log(result);
 *     },
 *     onConfirm: function (result) {
 *         console.log(result);
 *     },
 *     id: 'multiPickerBtn'
 * });
 *
 * @example
 * // 级联picker
 * weui.picker([
 * {
 *     label: '飞机票',
 *     value: 0,
 *     children: [
 *         {
 *             label: '经济舱',
 *             value: 1
 *         },
 *         {
 *             label: '商务舱',
 *             value: 2
 *         }
 *     ]
 * },
 * {
 *     label: '火车票',
 *     value: 1,
 *     children: [
 *         {
 *             label: '卧铺',
 *             value: 1,
 *             disabled: true // 不可用
 *         },
 *         {
 *             label: '坐票',
 *             value: 2
 *         },
 *         {
 *             label: '站票',
 *             value: 3
 *         }
 *     ]
 * },
 * {
 *     label: '汽车票',
 *     value: 3,
 *     children: [
 *         {
 *             label: '快班',
 *             value: 1
 *         },
 *         {
 *             label: '普通',
 *             value: 2
 *         }
 *     ]
 * }
 * ], {
 *    className: 'custom-classname',
 *    container: 'body',
 *    defaultValue: [1, 3],
 *    onChange: function (result) {
 *        console.log(result)
 *    },
 *    onConfirm: function (result) {
 *        console.log(result)
 *    },
 *    id: 'doubleLinePicker'
 * });
 */
function picker() {
    if (_sington) return _sington;

    // 配置项
    const options = arguments[arguments.length - 1];
    const defaults = $.extend({
        id: 'default',
        className: '',
        container: 'body',
        title: '',
        desc: '',
        onChange: $.noop,
        onConfirm: $.noop,
        onClose: $.noop
    }, options);

    // 数据处理
    let items;
    let isMulti = false; // 是否多列的类型
    if (arguments.length > 2) {
        let i = 0;
        items = [];
        while (i < arguments.length - 1) {
            items.push(arguments[i++]);
        }
        isMulti = true;
    } else {
        items = arguments[0];
    }

    // 获取缓存
    temp[defaults.id] = temp[defaults.id] || [];
    const result = [];
    const lineTemp = temp[defaults.id];
    const $picker = $($.render(pickerTpl, defaults));
    let depth = options.depth || (isMulti ? items.length : util.depthOf(items[0])), groups = '';

    // 显示与隐藏的方法
    function show() {
        $(defaults.container).append($picker);

        // 这里获取一下计算后的样式，强制触发渲染. fix IOS10下闪现的问题
        $.getStyle($picker[0], 'transform');

        //更改标题
        $picker.find('.weui-mask').addClass('weui-animate-fade-in');
        $picker.find('.weui-picker').addClass('weui-animate-slide-up');
    }

    function _hide(callback) {
        _hide = $.noop; // 防止二次调用导致报错

        $picker.find('.weui-mask').addClass('weui-animate-fade-out');
        $picker.find('.weui-picker')
            .addClass('weui-animate-slide-down')
            .on('animationend webkitAnimationEnd', function () {
                $picker.remove();
                _sington = false;
                defaults.onClose();
                callback && callback();
            });
    }

    function hide(callback) {
        _hide(callback);
    }

    // 初始化滚动的方法
    function scroll(items, level) {
        if (lineTemp[level] === undefined && defaults.defaultValue && defaults.defaultValue[level] !== undefined) {
            // 没有缓存选项，而且存在defaultValue
            const defaultVal = defaults.defaultValue[level];
            let index = 0, len = items.length;

            if (typeof items[index] == 'object') {
                for (; index < len; ++index) {
                    if (defaultVal == items[index].value) break;
                }
            } else {
                for (; index < len; ++index) {
                    if (defaultVal == items[index]) break;
                }
            }
            if (index < len) {
                lineTemp[level] = index;
            } else {
                console.warn('Picker has not match defaultValue: ' + defaultVal);
            }
        }
        $picker.find('.weui-picker__group').eq(level).scroll({
            items: items,
            temp: lineTemp[level],
            onChange: function (item, index) {
                //为当前的result赋值。
                if (item) {
                    result[level] = new Result(item);
                } else {
                    result[level] = null;
                }
                lineTemp[level] = index;

                if (isMulti) {
                    if (result.length == depth) {
                        defaults.onChange(result);
                    }
                } else {
                    /**
                     * @子列表处理
                     * 1. 在没有子列表，或者值列表的数组长度为0时，隐藏掉子列表。
                     * 2. 滑动之后发现重新有子列表时，再次显示子列表。
                     *
                     * @回调处理
                     * 1. 因为滑动实际上是一层一层传递的：父列表滚动完成之后，会call子列表的onChange，从而带动子列表的滑动。
                     * 2. 所以，使用者的传进来onChange回调应该在最后一个子列表滑动时再call
                     */
                    if (item.children && item.children.length > 0) {
                        $picker.find('.weui-picker__group').eq(level + 1).show();
                        !isMulti && scroll(item.children, level + 1); // 不是多列的情况下才继续处理children
                    } else {
                        //如果子列表test不通过，子孙列表都隐藏。
                        const $items = $picker.find('.weui-picker__group');
                        $items.forEach((ele, index) => {
                            if (index > level) {
                                $(ele).hide();
                            }
                        });

                        result.splice(level + 1);

                        defaults.onChange(result);
                    }
                }
            },
            onConfirm: defaults.onConfirm
        });
    }


    let _depth = depth;
    while (_depth--) {
        groups += groupTpl;
    }

    $picker.find('.weui-picker__bd').html(groups);
    show();

    if (isMulti) {
        items.forEach((item, index) => {
            scroll(item, index);
        });
    } else {
        scroll(items, 0);
    }

    $picker
        .on('click', '.weui-mask', function () {
            hide();
        })
        .on('click', '.weui-picker__btn', function () {
            hide();
        })
        .on('click', '#weui-picker-confirm', function () {
            defaults.onConfirm(result);
        });

    _sington = $picker[0];
    _sington.hide = hide;
    return _sington;
}

/**
 * datePicker 时间选择器，由picker拓展而来，提供年、月、日的选择。
 * @param options 配置项
 * @param {string=} [options.id=datePicker] 作为picker的唯一标识
 * @param {number=|string|Date} [options.start=2000] 起始年份，如果是 `Number` 类型，表示起始年份；如果是 `String` 类型，格式为 'YYYY-MM-DD'；如果是 `Date` 类型，就传一个 Date
 * @param {number=|string|Date} [options.end=2030] 结束年份，同上
 * @param {string=} [options.cron=* * *] cron 表达式，三位，分别是 dayOfMonth[1-31]，month[1-12] 和 dayOfWeek[0-6]（周日-周六）
 * @param {string=} [options.className] 自定义类名
 * @param {array=} [options.defaultValue] 默认选项的value数组, 如 [1991, 6, 9]
 * @param {function=} [options.onChange] 在picker选中的值发生变化的时候回调
 * @param {function=} [options.onConfirm] 在点击"确定"之后的回调。回调返回选中的结果(Array)，数组长度依赖于picker的层级。
 *
 *@example
 * // 示例1：
 * weui.datePicker({
 *     start: 1990,
 *     end: 2000,
 *     defaultValue: [1991, 6, 9],
 *     onChange: function(result){
 *         console.log(result);
 *     },
 *     onConfirm: function(result){
 *         console.log(result);
 *     },
 *     id: 'datePicker'
 * });
 *
 * // 示例2：
 * weui.datePicker({
 *      start: new Date(), // 从今天开始
 *      end: 2030,
 *      defaultValue: [2020, 6, 9],
 *      onChange: function(result){
 *          console.log(result);
 *      },
 *      onConfirm: function(result){
 *          console.log(result);
 *      },
 *      id: 'datePicker'
 *  });
 *
 *  // 示例3：
 * weui.datePicker({
 *      start: new Date(), // 从今天开始
 *      end: 2030,
 *      cron: '* * 0,6',  // 每逢周日、周六
 *      onChange: function(result){
 *          console.log(result);
 *      },
 *      onConfirm: function(result){
 *          console.log(result);
 *      },
 *      id: 'datePicker'
 *  });
 *
 *  // 示例4：
 * weui.datePicker({
 *      start: new Date(), // 从今天开始
 *      end: 2030,
 *      cron: '1-10 * *',  // 每月1日-10日
 *      onChange: function(result){
 *          console.log(result);
 *      },
 *      onConfirm: function(result){
 *          console.log(result);
 *      },
 *      id: 'datePicker'
 *  });
 */
function datePicker(options) {
    const nowDate = new Date();

    const defaults = $.extend({
        id: 'datePicker',
        onChange: $.noop,
        onConfirm: $.noop,
        start: nowDate.getFullYear() - 20,
        end: nowDate.getFullYear() + 20,
        defaultValue: [nowDate.getFullYear(), nowDate.getMonth() + 1, nowDate.getDate()],
        cron: '* * *'
    }, options);

    // 兼容原来的 start、end 传 Number 的用法
    if (typeof defaults.start === 'number') {
        defaults.start = new Date(`${defaults.start}/01/01`);
    } else if (typeof defaults.start === 'string') {
        defaults.start = new Date(defaults.start.replace(/-/g, '/'));
    }
    if (typeof defaults.end === 'number') {
        defaults.end = new Date(`${defaults.end}/12/31`);
    } else if (typeof defaults.end === 'string') {
        defaults.end = new Date(defaults.end.replace(/-/g, '/'));
    }

    const findBy = (array, key, value) => {
        for (let i = 0, len = array.length; i < len; i++) {
            const obj = array[i];
            if (obj[key] == value) {
                return obj;
            }
        }
    };

    const date = [];
    const interval = cron.parse(defaults.cron, defaults.start, defaults.end);
    let obj;
    do {
        obj = interval.next();

        const year = obj.value.getFullYear();
        const month = obj.value.getMonth() + 1;
        const day = obj.value.getDate();

        let Y = findBy(date, 'value', year);
        if (!Y) {
            Y = {
                label: year + '年',
                value: year,
                children: []
            };
            date.push(Y);
        }
        let M = findBy(Y.children, 'value', month);
        if (!M) {
            M = {
                label: month + '月',
                value: month,
                children: []
            };
            Y.children.push(M);
        }
        M.children.push({
            label: day + '日',
            value: day
        });
    }
    while (!obj.done);

    return picker(date, defaults);
}


/**
 * distPicker 多列选择器。
 * @param {Object} options 配置项
 * @param {number=} [options.depth] picker深度(也就是picker有多少列) 取值为1-3。如果为空，则取items第一项的深度。
 * @param {string=} [options.id=default] 作为picker的唯一标识，作用是以id缓存当时的选择。（当你想每次传入的defaultValue都是不一样时，可以使用不同的id区分）
 * @param {string=} [options.className] 自定义类名
 * @param {string=} [options.title] 自定义标题
 * @param {string=} [options.desc] 自定义描述
 * @param {string=} [options.container] 指定容器
 * @param {array=} [options.defaultValue] 默认选项的value数组
 * @param {function=} [options.onChange] 在picker选中的值发生变化的时候回调
 * @param {function=} [options.onConfirm] 在点击"确定"之后的回调。回调返回选中的结果(Array)，数组长度依赖于picker的层级。
 * @param {function=} [options.onClose] picker关闭后的回调
 *
 * @example
 * weui.distPicker({
 *       depth: 3,
 *       defaultValue: [110000, 110100, 110101],
 *       onChange: function (result) {
 *           console.log(result);
 *       },
 *       onConfirm: function (result) {
 *           console.log(result);
 *       },
 *       id: '地区选择器',
 *       className:"disPicker",
 *       desc:"省市区三级联动",
 *       title: '地区选择器'
 *   });
 * */
function distPicker() {
    const options = arguments[arguments.length - 1];
    const defaults = $.extend({
        id: 'default',
        className: '',
        container: 'body',
        title: '',
        desc: '',
        onChange: $.noop,
        onConfirm: $.noop,
        onClose: $.noop
    }, options);

    var dataArr = [];
    for (let i in districts['100000']) {
        let obj = {label: districts['100000'][i], value: i, children: []};
        for (let j in districts[i]) {
            let obj2 = {label: districts[i][j], value: j, children: []};

            for (let k in districts[j]) {
                let obj3 = {label: districts[j][k], value: k};
                obj2.children.push(obj3);
            }
            obj.children.push(obj2);
        }
        dataArr.push(obj);
    }
    return picker(dataArr, defaults);

}

export default {
    picker,
    datePicker,
    distPicker
};


example/example.js
// import 'weui';
import FastClick from 'fastclick';
import weui from '../src/zweui';

FastClick.attach(document.body);

/* dialog */
document.querySelector('#alertBtn').addEventListener('click', function () {
    weui.alert('自定义标题的alert', function () {
        console.log('ok');
    }, {
        title: '自定义标题'
    });
});


/* confirm */
document.querySelector('#confirmBtn').addEventListener('click', function () {
    weui.confirm('自定义标题的confirm', function () {
        console.log('yes');
    }, function () {
        console.log('no');
    }, {
        title: '自定义标题'
    });
});


/* toast */
document.querySelector('#toastBtn').addEventListener('click', function () {
    weui.toast('操作成功', {
        duration: 3000,
        className: 'bears'
    });
});


/* loading */
document.querySelector('#loadingBtn').addEventListener('click', function () {
    var loading = weui.loading('loading');
    setTimeout(function () {
        loading.hide();
    }, 3000);
});


/* actionSheet */
document.querySelector('#actionSheetBtn').addEventListener('click', function () {
    weui.actionSheet([
        {
            label: '拍照',
            onClick: function () {
                console.log('拍照');
            }
        }, {
            label: '从相册选择',
            onClick: function () {
                console.log('从相册选择');
            }
        }, {
            label: '其他',
            onClick: function () {
                console.log('其他');
            }
        }
    ], [
        {
            label: '取消',
            onClick: function () {
                console.log('取消');
            }
        }
    ], {
        title: 'actionTitle',
        className: 'custom-classname',
        onClose: function () {
            console.log('关闭');
        }
    });
});


/* topTips */
document.querySelector('#topTipsBtn').addEventListener('click', function () {
    weui.topTips('请填写正确的字段', {
        duration: 3000,
        className: 'custom-classname',
        callback: function () {
            console.log('close');
        }
    });
});


/* picker */
// 普通选择器
document.querySelector('#pickerBtn').addEventListener('click', function () {
    weui.picker([{
        label: '飞机票',
        value: 0
    }, {
        label: '火车票(disabled)',
        disabled: true,
        value: 1
    }, {
        label: '的士票',
        value: 2
    }, {
        label: '住宿费',
        value: 3
    }, {
        label: '礼品费(disabled)',
        value: 4,
        disabled: true,
    }, {
        label: '活动费',
        value: 5
    }, {
        label: '通讯费',
        value: 6
    }, {
        label: '补助',
        value: 7
    }, {
        label: '通讯费',
        value: 8
    }, {
        label: '其他',
        value: 9
    }], {
        defaultValue: [2],
        className: 'custom-classname',
        onChange: function (result) {
            //console.log(item, index);
            console.log(result);
        },
        onConfirm: function (result) {
            console.log('confirm');
            console.log(result);
        },
        id: 'picker',
        title: '单列选择器',
        desc: '单列选择器描述',
    });
});

// 时间选择器
document.querySelector('#datePickerBtn').addEventListener('click', function () {
    weui.datePicker({
        start: '2016-12-29',
        end: '2030-12-29',
        /**
         * https://zh.wikipedia.org/wiki/Cron
         * cron 表达式后三位
         * 示例：
         *  * * *                每天
         *  5 * *                每个月的5日
         *  1-10 * *             每个月的前10日
         *  1,5,10 * *           每个月的1号、5号、10号
         *  *\/2 * *             每个月的 1、3、5、7...日，注意写的时候斜杠“/”前面没有反斜杠“\”，这是因为是注释所以需要转义
         *  * 2 0                2月的每个周日
         *  * * 0,6              每个周末
         *  * * 3                每周三
         */
        cron: '* */2 0',
        defaultValue: [2017, 7, 9],
        onChange: function (result) {
            console.log(result);
        },
        onConfirm: function (result) {
            console.log(result);
        },
        id: 'datePicker',
        title: '日期选择器'
    });
});

// 多列选择器
document.querySelector('#multiPickerBtn').addEventListener('click', function () {
    weui.picker([
        {
            label: '1',
            value: '1'
        }, {
            label: '2',
            value: '2'
        }, {
            label: '3',
            value: '3'
        }
    ], [
        {
            label: 'A',
            value: 'A'
        }, {
            label: 'B',
            value: 'B'
        }, {
            label: 'C',
            value: 'C'
        }
    ], {
        defaultValue: ['3', 'A'],
        onChange: function (result) {
            console.log(result);
        },
        onConfirm: function (result) {
            console.log(result);
        },
        id: 'multiPickerBtn',
        onClose: function () {
            console.log('onClose');
        },
        title: '多列选择器'
    });
});

// 级联选择器
document.querySelector('#cascadePickerBtn').addEventListener('click', function () {
    weui.picker([
        {
            label: '广东',
            value: 0,
            children: [
                {
                    label: '广州',
                    value: 0,
                    children: [
                        {
                            label: '海珠',
                            value: 0
                        }, {
                            label: '番禺',
                            value: 1
                        }
                    ]
                }, {
                    label: '佛山',
                    value: 1,
                    children: [
                        {
                            label: '禅城',
                            value: 0
                        }, {
                            label: '南海',
                            value: 1
                        }
                    ]
                }
            ]
        }, {
            label: '广西',
            value: 1,
            children: [
                {
                    label: '南宁',
                    value: 0,
                    children: [
                        {
                            label: '青秀',
                            value: 0
                        }, {
                            label: '兴宁',
                            value: 1
                        }
                    ]
                }, {
                    label: '桂林',
                    value: 1,
                    children: [
                        {
                            label: '象山',
                            value: 0
                        }, {
                            label: '秀峰',
                            value: 1
                        }
                    ]
                }
            ]
        }
    ], {
        depth: 3,
        defaultValue: [0, 1, 1],
        onChange: function (result) {
            console.log(result);
        },
        onConfirm: function (result) {
            console.log(result);
        },
        id: 'cascadePicker2',
        title: '嵌套选择器'
    });
});

// 时间选择器
document.querySelector('#distPickerBtn').addEventListener('click', function () {
    weui.distPicker({
        depth: 3,
        defaultValue: [110000, 110100, 110101],
        onChange: function (result) {
            console.log(result);
        },
        onConfirm: function (result) {
            console.log(result);
        },
        id: 'distPicker',
        className: 'distPicker',
        desc: '省市区三级联动',
        title: '地区选择器'
    });
});

/* searchbar */
weui.searchBar('#searchBar');


/* slider 因为需要获取长度，所以必须要在slider显示的时候才调用weui.slider*/
var isSetSlider = false;

function setSlider() {
    if (isSetSlider) return;
    isSetSlider = true;

    // 普通slider
    var sliderValue = document.getElementById('sliderValue');
    weui.slider('#slider', {
        defaultValue: 50,
        onChange: function (percent) {
            sliderValue.innerHTML = Math.round(percent);
            console.log(percent);
        }
    });

    // 带step的slider
    var sliderStepValue = document.getElementById('sliderStepValue');
    weui.slider('#sliderStep', {
        step: 10,
        defaultValue: 40,
        onChange: function (percent) {
            sliderStepValue.innerHTML = Math.round(percent);
            console.log(percent);
        }
    });

    // 分块的slider
    var sliderBlockValue = document.getElementById('sliderBlockValue');
    weui.slider('#sliderBlock', {
        step: 100 / 3,
        defaultValue: 33.333,
        onChange: function (percent) {
            sliderBlockValue.innerHTML = Math.round(percent);
            console.log(percent);
        }
    });
}


/* tab */
weui.tab('#tab', {
    defaultIndex: 0,
    onChange: function (index) {
        console.log(index);

        if (index == 3) {
            setSlider(); // 设置slider
        }
    }
});


/* form */
// 约定正则
var regexp = {
    regexp: {
        IDNUM: /(?:^\d{15}$)|(?:^\d{18}$)|^\d{17}[\dXx]$/,
        VCODE: /^.{4}$/
    }
};

// 失去焦点时检测
weui.form.checkIfBlur('#form', regexp);

// 表单提交
document.querySelector('#formSubmitBtn').addEventListener('click', function () {
    weui.form.validate('#form', function (error) {
        console.log(error);
        if (!error) {
            var loading = weui.loading('提交中...');
            setTimeout(function () {
                loading.hide();
                weui.toast('提交成功', 3000);
            }, 1500);
        }
    }, regexp);
});


/* 图片自动上传 */
var uploadCount = 0, uploadList = [];
var uploadCountDom = document.getElementById('uploadCount');
weui.uploader('#uploader', {
    url: 'http://' + location.hostname + ':8002/upload',
    auto: true,
    type: 'file',
    fileVal: 'fileVal',
    compress: {
        width: 1600,
        height: 1600,
        quality: .8
    },
    onBeforeQueued: function (files) {
        if (['image/jpg', 'image/jpeg', 'image/png', 'image/gif'].indexOf(this.type) < 0) {
            weui.alert('请上传图片');
            return false;
        }
        if (this.size > 10 * 1024 * 1024) {
            weui.alert('请上传不超过10M的图片');
            return false;
        }
        if (files.length > 5) { // 防止一下子选中过多文件
            weui.alert('最多只能上传5张图片，请重新选择');
            return false;
        }
        if (uploadCount + 1 > 5) {
            weui.alert('最多只能上传5张图片');
            return false;
        }

        ++uploadCount;
        uploadCountDom.innerHTML = uploadCount;
    },
    onQueued: function () {
        uploadList.push(this);
        console.log(this);
    },
    onBeforeSend: function (data, headers) {
        console.log(this, data, headers);
        // $.extend(data, { test: 1 }); // 可以扩展此对象来控制上传参数
        // $.extend(headers, { Origin: 'http://127.0.0.1' }); // 可以扩展此对象来控制上传头部

        // return false; // 阻止文件上传
    },
    onProgress: function (procent) {
        console.log(this, procent);
    },
    onSuccess: function (ret) {
        console.log(this, ret);
    },
    onError: function (err) {
        console.log(this, err);
    }
});

// 缩略图预览
document.querySelector('#uploaderFiles').addEventListener('click', function (e) {
    var target = e.target;

    while (!target.classList.contains('weui-uploader__file') && target) {
        target = target.parentNode;
    }
    if (!target) return;

    var url = target.getAttribute('style') || '';
    var id = target.getAttribute('data-id');

    if (url) {
        url = url.match(/url\((.*?)\)/)[1].replace(/"/g, '');
    }
    var gallery = weui.gallery(url, {
        className: 'custom-name',
        onDelete: function () {
            weui.confirm('确定删除该图片？', function () {
                --uploadCount;
                uploadCountDom.innerHTML = uploadCount;


                for (var i = 0, len = uploadList.length; i < len; ++i) {
                    var file = uploadList[i];
                    if (file.id == id) {
                        file.stop();
                        break;
                    }
                }
                target.remove();
                gallery.hide();
            });
        }
    });
});


/* 图片手动上传 */
var uploadCustomFileList = [];

// 这里是简单的调用，其余api请参考文档
weui.uploader('#uploaderCustom', {
    url: 'http://localhost:8002/upload',
    auto: false,
    onQueued: function () {
        uploadCustomFileList.push(this);
    }
});

// 手动上传按钮
document.getElementById('uploaderCustomBtn').addEventListener('click', function () {
    uploadCustomFileList.forEach(function (file) {
        file.upload();
    });
});

// 缩略图预览
document.querySelector('#uploaderCustomFiles').addEventListener('click', function (e) {
    var target = e.target;

    while (!target.classList.contains('weui-uploader__file') && target) {
        target = target.parentNode;
    }
    if (!target) return;

    var url = target.getAttribute('style') || '';
    var id = target.getAttribute('data-id');

    if (url) {
        url = url.match(/url\((.*?)\)/)[1].replace(/"/g, '');
    }
    var gallery = weui.gallery(url, {
        onDelete: function () {
            weui.confirm('确定删除该图片？', function () {
                var index;
                for (var i = 0, len = uploadCustomFileList.length; i < len; ++i) {
                    var file = uploadCustomFileList[i];
                    if (file.id == id) {
                        index = i;
                        break;
                    }
                }
                if (index !== undefined) uploadCustomFileList.splice(index, 1);

                target.remove();
                gallery.hide();
            });
        }
    });
});
